### ISSUE TITLE
Pushing image from Windows 2016 to DTR: invalid manifest

### ISSUE THREAD

**Issue by SjaakS** [docker/for-win#202]
Hi,
I am trying to push images from a Windows 2016 machine to the trusted registry. Pushing from a linux machine works fine, but pushing from the Windows machine throws an exception.

UCP installed on a Ubuntu 14.04.5 LTS machine:

docker version
Client:
Version: 1.12.2-cs2
API version: 1.24
Go version: go1.6.3
Git commit: d5fda6e
Built: Thu Oct 13 01:58:32 2016
OS/Arch: linux/amd64
Server:
Version: 1.12.2-cs2
API version: 1.24
Go version: go1.6.3
Git commit: d5fda6e
Built: Thu Oct 13 01:58:32 2016
OS/Arch: linux/amd64

DTR installed on a Ubuntu 14.04.5 LTS machine:

docker version
Client:
Version: 1.12.2-cs2
API version: 1.24
Go version: go1.6.3
Git commit: d5fda6e
Built: Thu Oct 13 01:58:32 2016
OS/Arch: linux/amd64
Server:
Version: 1.12.2-cs2
API version: 1.24
Go version: go1.6.3
Git commit: d5fda6e
Built: Thu Oct 13 01:58:32 2016
OS/Arch: linux/amd64

Expected behavior

According to the presentations it should be possible to push images from a linux- and a Windows 2016 machine

Actual behavior

Pushing a linux based image from a linux machine

docker version
Client:
Version: 1.12.2-cs2
API version: 1.24
Go version: go1.6.3
Git commit: d5fda6e
Built: Thu Oct 13 01:58:32 2016
OS/Arch: linux/amd64
Server:
Version: 1.12.2-cs2
API version: 1.24
Go version: go1.6.3
Git commit: d5fda6e
Built: Thu Oct 13 01:58:32 2016
OS/Arch: linux/amd64

docker login -u .. -p .. dtr
docker push dtr/...
works fine.

Pushing a windows nanoserver image from a Windows 2016 server

docker version
Client:
Version: 1.12.2-cs2-ws-beta
API version: 1.25
Go version: go1.7.1
Git commit: 050b611
Built: Thu Oct 11 02:35:40 2016
OS/Arch: windows/amd64
Server:
Version: 1.12.2-cs2-ws-beta
API version: 1.25
Go version: go1.7.1
Git commit: 050b611
Built: Thu Oct 11 02:35:40 2016
OS/Arch: windows/amd64

docker login -u .. -p .. dtr
docker push dtr/...
throws the following error:
The push refers to a repository [...]
2c195a33d84d: Preparing
342d4e407550: Preparing
342d4e407550: Skipped foreign layer
2c195a33d84d: Skipped foreign layer
docker : manifest invlaid: manifest invalid
At line:1 char:1
+ docker push .....
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo : NotSpecified: (manifest invalid: manifest invalid:Stri
ng) [], RemoteException
+ FullQualifiedErrorId : NativeCommandError

**Comment by rneugeba** [comment]
I think it's currently not possible to push windows container images to a private registry. @friism could you clarify?

**Comment by friism** [comment]
Pushing to a private repository works, I believe, but there's something going on with DTR.

Ping @pdevine @and @dhiltgen

**Comment by dhiltgen** [comment]
I think I saw chatter on slack that the problem is DTR assumes all layers are present, but apparently there's some "magic" in the stack that prevents the msft base layers from being pushed/pulled in certain circumstances.  Fwiw, I think this is a really bad idea and is going to lead to a very poor UX for on-prem customers, but I'm guessing it's a contractual issue.  Hopefully we can convince msft this is a bad idea.

Long story short, please file an issue on dhe-deploy to track. (maybe there's already one... not sure)

**Comment by arthi-ramachandran** [comment]
@friism , having same issue abt pushing a windows server 2016 image to DTR. Can you please advise whether this is fixed ?

**Comment by friism** [comment]
@arthi-ramachandran @SjaakS if you `docker save -o foo.tar.gz` and then `docker load`, I think you should be able to push the resulting image. Let me know if it works.

**Comment by arthi-ramachandran** [comment]
@friism it seems this method is not going via the DPR. Instead directly saving and scp of the image.

**Comment by volak** [comment]
Hi all - I just ran into this issue today.  Pushing my windows docker images to a linux registry is not working - after much fiddling I finally reached this error.  Using the guide [here](https://www.digitalocean.com/community/tutorials/how-to-set-up-a-private-docker-registry-on-ubuntu-14-04) I setup a private repo with ssl only to continually receive ` x509: certificate signed by unknown authority` even though `curl https://my-host/v2/` worked fine.

It was only after tagging my images `my-host:443/project` that I was able to push at all, and now I'm receiving this error.

Any updates on workarounds?

**Comment by friism** [comment]
@volak I posted the workaround here: https://github.com/docker/docker/issues/27580#issuecomment-262346159

**Comment by volak** [comment]
@friism Thanks for the update.. that process is .. unfortunate.  I'll have to transfer around 6gb packages to deploy each of my docker containers.  Not to mention the editing process is not really something I can put into ansible.

If I was running the registry on windows would I have this problem?  Is running registry on windows even possible?

**Comment by friism** [comment]
@volak you can [run the registry in Windows](https://hub.docker.com/r/sixeyed/registry/) but it has the same behavior

**Comment by volak** [comment]
@friism You mention in a previous comment https://github.com/docker/docker/issues/27580#issuecomment-262162433 that the images would be pushable if docker could be forced to push foreign layers?  Is there a custom build or repo for that build?  I could take a stab at overriding "skip foreign layers" but Go is not my favorite language..

**Comment by friism** [comment]
@volak no, currently the only approach is what's outlined in that comment

**Comment by friism** [comment]
I'm closing this issue since it's related to DTR and `docker/docker`.

**Comment by volak** [comment]
Sorry - is there an open ticket somewhere else I can track the progress of this?

unless I'm mistaken until this is fixed there is no way to push images from a build server into production (without manual workaround) - which is a blocking issue for us

**Comment by friism** [comment]
Here's the issue: https://github.com/docker/docker/issues/27580

I'll add note asking for it to be re-opened.

**Comment by muditmn** [comment]
i want to pull image from a DTR but my windows node runs either internet or DTR at a single time. can anybody help me with that.
SYSTEM OS:- Windows server 2016
